### PR TITLE
[GKD] interpolate in prob. space

### LINKED ISSUE #2202: GKD generalized JSD implementation question
Hi, in trl GKD generalized JSD implementation we saw log prob mixture of teacher and student, while in the original paper it seems to mix the prob before applying log https://github.com/huggingface/trl/blob/main/trl/trainer/gkd_trainer.py#L186-L188
is this discrepancy a known improvement to the original paper? it really changes the behavior: say beta=0.5 by default, and student has a tiny prob close to 0 on a logit while teacher has a prob nearly 1, when mixing prob, the interpolated value will be about 0.5, but when mixing log prob, log prob of the tiny value will dominate and the interpolated value will be a large negative.

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

fix the JSD by interpolating in the prob. space

fixes #2202

### CODE REVIEW

File: trl/trainer/gkd_trainer.py [hunk 1]
```diff
@@ -179,17 +179,21 @@ def generalized_jsd_loss(
         student_logits = student_logits / temperature
         teacher_logits = teacher_logits / temperature
 
+        # Compute the interpolated probabilities
+        student_probs = F.softmax(student_logits, dim=-1)
+        teacher_probs = F.softmax(teacher_logits, dim=-1)
+        mixture_probs = beta * student_probs + (1 - beta) * teacher_probs
+        # Compute the log of the mixture distribution
+        mixture_log_prob = mixture_probs.log()
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py:187]: probs are around 1e-20 on average, don't you think we can have numerical issues here?

Maybe keeping log-probs is safer:

```python
# log(a + b) = log(exp(log(a)) + exp(log(b))) -> for mixture
mixture_log_probs = torch.logsumexp(torch.stack([
    log_student_probs + torch.log(torch.tensor(beta)),
    log_teacher_probs + torch.log(torch.tensor(1 - beta))
]), dim=0)
```

File: trl/trainer/gkd_trainer.py [hunk 2]
```diff
@@ -183,13 +183,22 @@ def generalized_jsd_loss(
...
+        # log(a + b) = log(exp(log(a)) + exp(log(b))) -> for mixture
+        mixture_log_probs = torch.logsumexp(
+            torch.stack(
+                [
+                    student_log_probs + torch.log(torch.tensor(beta, dtype=student_log_probs.dtype)),
+                    teacher_log_probs + torch.log(torch.tensor(1 - beta, dtype=teacher_log_probs.dtype)),
+                ]
+            ),
+            dim=0,
+        )
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py:196]:
```suggestion
        beta = torch.tensor(beta, dtype=student_log_probs.dtype)
        mixture_log_probs = torch.logsumexp(
            torch.stack([student_log_probs + torch.log(beta), teacher_log_probs + torch.log(1 - beta)])
        )
```

not sure about the dim etc, you'll need to double check

Comment (kashif) [reply trl/trainer/gkd_trainer.py:196]: i checked and its needed

**Review by qgallouedec — APPROVED** [review]
LGTM! Could be nice to have ref results, if you've time
